url47af6b1535_rolandotr_db_comparison_tgz-pJ8-methodology_HistoryJ8/src/utils/SpiderChart.java

```diff
@@ -0,0 +1,273 @@
+		tree.setNumberOfRounds(128);
+		DBProtocol poulidor = new PoulidorProtocol();
+		poulidor.setNumberOfRounds(128);
+		DBProtocol tma = new TMAProtocol();
+		tma.setNumberOfRounds(128);
+		DBProtocol[] firstSet = new DBProtocol[]{
+				best,tree,poulidor,tma
+		};
+		Attribute[] attributes = new Attribute[]{
+				new MafiaFraudProbability(new ProbabilityRelation(), new LogScale(2)),
+				new DistanceFraudProbability(new ProbabilityRelation(), new LogScale(2)),
+				new TerroristFraudProbability(new ProbabilityRelation(), new LogScale(2)),
+				new Memory(new MemoryRelation(), new KbitsScale()),
+		};
+
+		printSpiderChart(firstSet, attributes, "tree_poulidor_tma.tex");
+	}
+	
+	public static void printSpiderChart(DBProtocol[] protocols, Attribute[] attributes, 
+			String fileName) throws IOException{
+		FileWriter writer = new FileWriter(fileName);
+		//first, we print the header
+		writer.append("\\begin{tikzpicture}[label distance=.15cm,rotate=30,scale=0.5]"+newLine);
+		writer.append("\\tkzKiviatDiagram{");
+		addAttributes(attributes, writer);
+		writer.append("}"+newLine);
+		double[][] normalizedAttributes = normalizeAttributes(protocols, attributes);		
+		for (int i = 0; i < protocols.length; i++) {
+			addProtocol(i, attributes, normalizedAttributes, writer);
+		}
+	    writer.append("\\end{tikzpicture}");
+		writer.close();	
+	}
+
+	private static double[][] normalizeAttributes(DBProtocol[] protocols, 
+			Attribute[] allAttributes) {
+		double[][] result = new double[allAttributes.length][protocols.length];
+		for (int i = 0; i < allAttributes.length; i++) {
+			result[i] = normalizeAttributes(protocols, 
+					allAttributes[i]);
+		}
+		return result;
+	}
+
+	private static double[] normalizeAttributes(DBProtocol[] protocols,
+			Attribute attribute) {
+		if (attribute.getValue() instanceof Double){
+			return normalizeDoubleAttributes(protocols,attribute);
+		}
+		else if (attribute.getValue() instanceof Integer){
+			return normalizeIntegerAttributes(protocols,attribute);
+		}
+		else if (attribute.getValue() instanceof Long){
+			return normalizeLongAttributes(protocols,attribute);
+		}
+		else if (attribute.getValue() instanceof Boolean){
+			return normalizeBooleanAttributes(protocols,attribute);
+		}
+		else{
+			throw new RuntimeException();
+		}
+	}
+
+	private static double[] normalizeDoubleAttributes(DBProtocol[] protocols,
+			Attribute attribute) {
+		double max = 0;
+		double min = Double.MAX_VALUE;
+		for (int i = 0; i < protocols.length; i++) {
+			double tmp = (Double)protocols[i].getAttribute(attribute).getScaledValue();
+			if (tmp > max) max = tmp;
+			if (tmp < min) min = tmp;
+		}
+		double[] result = new double[protocols.length];
+		for (int i = 0; i < result.length; i++) {
+			double tmp = (Double)protocols[i].getAttribute(attribute).getScaledValue();
+			if (max == min) result[i] = 1;
+			else result[i] = 1-(tmp-min)/(max-min);
+		}
+		return result;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 174:187, 192:205, 209:222.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.